Topic training (#10)
WisconsinRobotics/WRoverPlayground: src/topic_training/src/Begin_Here.md

# WRoverPlayground - Topic Training
Now that you have completed the Tools Training, you are ready to learn about ROS topics. From the Setup Training, you should be familiar with:
- the basics of ROS Nodes

> - rosnode
> - and roscore.

Reviewer: Here if you are referring to the commands `rosnode` and `roscore`, use \`\` around them so there's less ambiguity.
```suggestion
- `rosnode`
- and `roscore`.
```

Author: Fixed in 7ece46c.

Once you complete this training, you should be familiar with:
- the purpose of topics
- the relationship between Publishers and Subscribers
- setting up Publishers and Subscribers
- running Publishers and Subscribers
- rqt_graph

> - rqt_plot

Reviewer: Not sure if you and Ben already talked about this, but I thought `rqt_graph` and `rqt_plot` should be covered in debug toolings training.

Author: Fixed in abd4191.

- custom topic types
- and rostopic type {topic}.

## The Purpose of Topics

> The word "topic" sounds very general, but it is actually a specific concept that allows ROS nodes to communicate with each other. This is extremely important because one node cannot possibly contain all the information a rover would need.

Reviewer 2: It could, is the thing.  This is typically called a 'monolith', and becomes very hard to maintain without extreme diligence.  They are often much more complex as well.

I would argue that the topic is a way to stream data out of or into a program.

Reviewer: Agree with Ben's comments on this paragraph. For this part maybe it's better to highlight that we need a way for nodes to communicate and share data with each other, since we put things in different classes to avoid one huge class, like the monolith Ben talked about. For instance, the node that handles controller input would need some way to communicate with the node that manages the drivetrain.

Author: Fixed in f580517.

Reviewer 2: I think this a very good improvement and is a good view of topics.  Nice job!

> Topics allow us to specialize nodes, so they can execute their own code/tasks, while also sending and receiving information. For example, a node could receive data and send it to another one, a node could receive and analyze data and have another node do some action with it, a node could receive data from multiple nodes, etc.

Reviewer 2: I like the comment about the specialization of nodes.  I would expound on it and additionally emphasize the 'hot-swap'-ability of code using topics.  Since topics don't really care who is publishing or subscribing, it's not only easy to write a node that serves, say, GPS data from a physical sensor, but it is equally easy to create a mock node to test the other nodes without hardware *without changing the other nodes*, since the interface provided through the published data is equivalent.

Author: Fixed in f580517.

## Publisher/Subscriber Relationship

> The Publisher and Subscriber are both nodes that are connected by a topic, allowing them to send messages between themselves. The Publisher node publishes messages on the topic, while the Subscriber node subscribes to the same topic to receive the Publisher's messages. 

Reviewer 2: Nit-pick; don't use a topic to send data to yourself (comparatively *very* inefficient). 
```suggestion
The Publisher and Subscriber are both nodes that are connected by a topic, allowing them to send messages between each other. The Publisher node publishes messages on the topic, while the Subscriber node subscribes to the same topic to receive the Publisher's messages. 
```

Author: Fixed in 3e8569b.

> The Publisher and Subscriber are both nodes that are connected by a topic, allowing them to send messages between themselves. The Publisher node publishes messages on the topic, while the Subscriber node subscribes to the same topic to receive the Publisher's messages. 

Reviewer: Would be helpful to also add that a topic can have multiple subscribers, allowing a single node to communicate with multiple other nodes. 

I guess you could also have multiple publishers on a single topic, but seems like a very corner case and doesn't need to be mentioned here.

Author: Fixed in d7207f7.

> Publishers and Subscribers do not need to be in separate files. In fact, if the Publisher and Subscriber rely on each other's data, it may be better to keep them in the same file. An example of the Publisher and Subscriber residing in the same file will be provided in the next section.

Reviewer 2: See nit-pick comment, I would avoid suggesting publishing to oneself.

Author: Fixed in d7207f7.

## Setting up a Publisher and Subscriber
Within this directory, you will find two files called Demo(.cpp or .py). These files will show a basic class containing both a Publisher and Subscriber. There will be comments throughout the files that explain the purpose of specific sections of the code.

Note that there are quite a few sections of code that are marked as "Optional." You do not need to include everything in the examples when creating your own code. They are simply showing one way of setting up a Publisher and Subscriber in the same file.

> One file will be written in C++, and the other will be in Python. I recommend discussing with one of the Software Leads to determine which part of the rover you would like to be assigned to. This will help you decide which language to focus on for this training. However, being familiar with both languages will not hurt.

Reviewer 2: I would encourage using the language they know rather than the language they're going to use.  Most of the students will have broad deference for what they're working on, so assigning them a project/language is not very efficient.

Author: Fixed in 564eefa.

Go through one of the files, making sure you understand all the code and comments. Come back to this file once you feel ready to move on.

## Running a Publisher and Subscriber

> Before you are able to run the example code, find the CMakeLists.txt file that is outside of the src directory but still inside of the topic_training directory. 

Reviewer: @bennowotny Ben let me know your opinion on this, but I feel like at this point it's probably better to not have new members touch `CMakeLists` at all, in case it adds to confusion or just seems daunting in general. I recommend just leaving the code block below uncommented in `CMakeLists` to start with instead of having new members do it themselves.

Reviewer 2: Yes, I would avoid this.  We are probably going to have the custom types be pre-built and just show them how it's done/direct them to relevant code to show them how it would be done.

Author: Fixed in ead3ab7.

Author: Fixed in ead3ab7.

This file is mostly commented out. Make sure these are uncommented:

- find_package(catkin REQUIRED
    roscpp
    rospy
    std_msgs
    message_generation
  )

- catkin_package(
    CATKIN_DEPENDS message_runtime
  )

- include_directories(
    ${catkin_INCLUDE_DIRS}

>   )

Reviewer 2: Tag these blocks with comments in CMake to let the students know they're looking in the right spot.

Author: Since I removed the unnecessary comments, do I still need to tag the used blocks? I feel like it will be simpler to find the right spot now.

Reviewer 2: You probably don't need to.  It may be helpful, but not required.

The C++ and the Python file both require these to work.

If you chose C++, also uncomment these:

- add_executable(Demo src/Demo.cpp)
  add_executable(PlotDemo src/PlotDemo.cpp)
  add_executable(MotorChallenge src/MotorChallenge.cpp)

- add_dependencies(Demo 
    ${${PROJECT_NAME}_EXPORTED_TARGETS} 
    ${catkin_EXPORTED_TARGETS}
  )
  add_dependencies(PlotDemo 
    ${${PROJECT_NAME}_EXPORTED_TARGETS} 
    ${catkin_EXPORTED_TARGETS}
  )
  add_dependencies(MotorChallenge 
    ${${PROJECT_NAME}_EXPORTED_TARGETS} 
    ${catkin_EXPORTED_TARGETS}
  )

- target_link_libraries(Demo 
    ${catkin_LIBRARIES}
  )
  target_link_libraries(PlotDemo
    ${catkin_LIBRARIES}
  )
  target_link_libraries(MotorChallenge
    ${catkin_LIBRARIES}
  )

If you chose Python, also uncomment this:

- catkin_install_python(PROGRAMS src/Demo.py src/PlotDemo.py src/MotorChallenge.py
    DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}

>   )

Reviewer 2: You could uncomment all of these.  CMake is complicated at a first pass, so I would avoid making large, unexplained changes like these.  Maybe a reference or explanation?  Otherwise, avoid this kind of change altogether.

Author: Fixed in 08a9c40.

In future projects, you might need to use some of the other configurations, builds, and installs, but for the purposes of this training, you only need these for the example code to run.

> You can also check out the package.xml file, which you also may need to tailor for future projects. However, everything should already be set up correctly, so do not change anything.

Reviewer 2: Maybe highlight the parts you need or why you need them.  If not, include a reference to the package.xml file.

Author: Fixed in ead3ab7.

> In the terminal, run the command *roscore*. In a new terminal window, cd into the catkin workspace. In this case, use *cd WRoverPlayground*. Then, use the commands *catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=ON* and *source devel/setup.bash*. These commands adjust the catkin workspace based on the CMakeLists.txt and package.xml files. You must run these commands everytime you change these two files. 

Reviewer 2: For code or commands, use literal markers like `this` (\`this\`).  This comment applies to all code examples or commands in this file.

Author: Fixed in 7ece46c.

> In the terminal, run the command *roscore*. In a new terminal window, cd into the catkin workspace. In this case, use *cd WRoverPlayground*. Then, use the commands *catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=ON* and *source devel/setup.bash*. These commands adjust the catkin workspace based on the CMakeLists.txt and package.xml files. You must run these commands everytime you change these two files. 

Reviewer 2: For ordered instructions, use an ordered list:
1. Something to do first
2. Something to do second
3. Extra things to do afterwards

Author: Fixed in bf4bc70.

> In the terminal, run the command *roscore*. In a new terminal window, cd into the catkin workspace. In this case, use *cd WRoverPlayground*. Then, use the commands *catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=ON* and *source devel/setup.bash*. These commands adjust the catkin workspace based on the CMakeLists.txt and package.xml files. You must run these commands everytime you change these two files. 

Reviewer: To add to Ben's comments, avoid language such as "cd into the catkin workspace" as there's a good chance new members won't know what cd means. With each command remember to explain what it does.

Author: Fixed in 475a96b.

> With everything hopefully set up properly, you may now run *rosrun topic_training Demo* or *rosrun topic_training Demo.py*. In the terminal, you should see the program continually print "hello world {count} I heard: [hello world {count}]" until the terminal receives Ctrl-C or the program crashes.

Reviewer: Explain that `rosrun topic_training Demo` is for the C++ version and `rosrun topic_training Demo.py` is for Python.

Also remember to explain what control C does.
 
```suggestion
With everything hopefully set up properly, you may now run `rosrun topic_training Demo` to execute the C++ version of the demo, or `rosrun topic_training Demo.py` to execute the Python version of the demo. In the terminal, you should see the program continually print "hello world {count} I heard: [hello world {count}]" until you exit the program by inputting `Ctrl + c` on the terminal or the program crashes.
```

Author: Fixed in b79b02d.

If you run rqt_graph on the PlotDemo(.cpp or .py) file, it should show just the name of the file (+ the ID number if you are using Python) by itself with no arrow. If you run rqt_graph when both files are running, there might be an error in the terminal, but the graph application should show both files. PlotDemo(.cpp or .py) will be by itself with no arrow, and Demo(.cpp or .py) will be by itself with or without an arrow depending on the order you executed the two files.

Feel free to look through the file PlotDemo(.cpp or .py) to see how the math works and to see another example of a Publisher but with a different message type. For simplicity's sake, the comments in these files are only explaining the math since you should already understand the basic ROS functions from the previous section of this training.

> 

Reviewer 2: This feels like it's part of the tools training. The nodes could print, maybe just use rqt_* to reinforce how these concepts are toolable.

Author: Fixed in abd4191

> ### Defining a Custom Topic Type

Reviewer 2: You need a custom topic type for the messages?  You should probably have 1 custom message type just to show it.

Author: Fixed in 661fa32.

A topic's type is determined by the type of the message being published. If you run the command *rostopic type chatter*, you should see the terminal print out "std_msgs/String" while roscore and rosrun are executing Demo(.cpp or .py). The command will print out "std_msgs/Float64" if you are executing PlotDemo(.cpp or .py).

> A topic can have messages of type int8, int16, int32, int64, uint*, float32, float64, String, time, duration, other msg files, and array[] of variable length or fixed length. To message a different type, just change where ever there would be a String in Demo(.cpp or .py) to your desired type and update the includes (C++) or imports (Python) at the top of the file.

Reviewer: Could add a short sentence or paragraph at the end of this segment to remind students to use message types instead of variable types. For instance using `std_msgs/String` instead of `String`.

Author: Fixed in a8f38fc

Author: Fixed in a8f38fc.

## Challenge

> Once you have completed the training from WRoverPlayground, you should have some working code that will be put into a simulation. The simulation will have a rover that your code is meant to control its drive through mock navigation data to reach some target/beacon. 

Reviewer: Wording is a bit confusing. 

```suggestion
Once you have completed the training from WRoverPlayground, you should have some working code that will be put into a simulation. The simulation will include a rover that your code would need to control using mock navigation data to reach targets/beacons. 
```

Author: Fixed in 239cba0.

> You will take in sensor data as a 180-element array of floats that represents a 180 degree view of the robot. Each element is a degree-sized slice of the robot's view. One element will contain a float, while the other elements will be undefined. The float in this array represents the distance from the head of the robot to the head of the target. The location of the element in the array represents its position relative to the robot. If the float is in the 90th element, the target is straight ahead. If the float is before the 90th element, the target is to the robot's left, and vice versa. 

Reviewer: The wording of "one element" might be a bit misleading as I'm under the impression that there will be multiple beacons. Might be better to switch to "some elements" instead.

```suggestion
You will take in sensor data as a 180-element array of floats that represents a 180 degree view of the robot. Each element is a degree-sized slice of the robot's view. Some elements will contain a float, while others will be undefined. The float in this array represents the distance from the head of the robot to the head of the target. The location of the element in the array represents its position relative to the robot. If the float is in the 90th element, the target is straight ahead. If the float is before the 90th element, the target is to the robot's left, and vice versa. 
```

Author: Fixed in bb4e4e0.

> You will take in sensor data as a 180-element array of floats that represents a 180 degree view of the robot. Each element is a degree-sized slice of the robot's view. One element will contain a float, while the other elements will be undefined. The float in this array represents the distance from the head of the robot to the head of the target. The location of the element in the array represents its position relative to the robot. If the float is in the 90th element, the target is straight ahead. If the float is before the 90th element, the target is to the robot's left, and vice versa. 

Reviewer: I feel like this paragraph needs to be further expanded in general. There's not enough context for new members to understand what the array represents and how to interpret it. You can mention things like:
- This array is supposed to mock the input from a LIDAR which is what we use for our competition robot to add more context. Maybe even include a link to a webpage that explains what a LIDAR is. 
- The array is a series of readings of distances detected by the LIDAR. It is clockwise, with the 0th index mapping to 9 O'Clock, the 89th index mapping to 12 O'Clock, and the 189th index mapping to 3 O'Clock.
- A reading of `undefined` means that there are no obstacles detected in that direction (although from my memory the real LIDAR uses `Inf` instead of `undefined`).

Author: Fixed in 68e6610.

Using this information, you will output speeds, on the interval [-1, 1], for the robot until it reaches its target. There will be two separate values controlling the speed of the left side and the speed of right side. To learn more about how the controllers are divided between the left and right side, refer back to the general description of the training.

> Go into the MotorChallenge(.cpp or .py) file. This is where you will be writing your code. Feel free to use the examples provided in this training and any other resources that will be useful. The CMakeLists.txt file should already be set up to run your code if you uncommented correctly.

Reviewer 2: I would walk through more steps for how they're supposed to do this.

Reviewer 2: Maybe a picture, especially for the sensor.

Author: Fixed in 0f4ab2b.

- the basics of ROS Nodes
- `rosnode`
- `roscore`
- `rqt_graph`
- `rqt_plot`.

Once you complete this training, you should be familiar with:
- the purpose of topics
- the relationship between Publishers and Subscribers
- setting up Publishers and Subscribers
- running Publishers and Subscribers
- custom topic types
- and `rostopic type {topic}`.

## The Purpose of Topics
Though possible, it is not feasible to have one node contain all the information a rover would need. We need to spread out the information into different files. This is where ROS topics come in. The word "topic" sounds very general, but it is actually a specific concept that allows ROS nodes to communicate with each other. They are a way to stream data in and out of a program.

Using topics, our code now becomes 'hot-swappable.' Topics do not care which node is sending or receiving data, so we can easily create a node that serves data from a physical sensor as well as a mock node to test other nodes without needing to change those nodes. The reason is because the interfaces provided through the published data versus an actual sensor are equivalent.

> This allows us to specialize nodes, so they can execute their own code/tasks, while also sending and receiving information. For example, a node could receive data and send it to another one, a node could receive and analyze data and have another node do some action with it, a node could send data to multiple nodes, etc.

Reviewer 2: Maybe give a specific example, to make it concrete?

Author: @bennowotny Should this just be a real-world example, like USBs?

Reviewer 2: Maybe something we use on the robot, like the specialization of the motor driver programs being separated from how we drive the motors which is separated from how we gather user input.

Author: Fixed in f8df283.

## Publisher/Subscriber Relationship
The Publisher and Subscriber are both nodes that are connected by a topic, allowing them to send messages between each other. The Publisher node publishes messages on the topic, while the Subscriber node subscribes to the same topic to receive the Publisher's messages. 

Topics do not need to have a one-to-one relationship. For example, it is possible for a topic to have multiple Subscribers, which allows a single node to communicate with other nodes.

> Also, Publishers and Subscribers do not need to be in separate files. In fact, if the Publisher and Subscriber rely on each other's data, it may be better to keep them in the same file. However, they should not be on the same topic as this would be incredibly inefficient. A way to do this would be to have the Subscriber receive data from one topic, the data to be manipulated in some way, and the Publisher to send the result on another topic. 

Reviewer 2: Rather than 'be on the same topic', I would phrase the thing to avoid as 'a node publishing to itself'

Author: Fixed in e78d3b3

Publishing a message on a topic would look something like this:

``` Python
pi_str = "currently at %s" % pi
pub.publish(hello_str)
```

To break this down,

``` Python
pi_str = "currently at %s" % pi
```
Create the message.

``` Python
pub.publish(hello_str)
```
Publish the message using the previously initialized Publisher `pub`.

If you have glanced at the C++ example code, you may have noticed the function `ros::spinOnce()` and its differences with `ros::spin()`. Python only has `rospy.spin()`, which can be completely replaced by a for loop, which is why you do not see it in the Python example code.

## Running a Publisher and Subscriber
Before you are able to run the example code, find the CMakeLists.txt file that is outside of the src directory but still inside of the topic_training directory. This file will include the the configurations, builds, and installs required to run your program as well as the files you will be running. Do not change anything in the file as it should already be set up for the demo files to work. 

> The `find_package`, `catkin_package`, and `include_directories` specifications are required for both the C++ and Python files to work. These provide the packages and directories needed for running ROS and ROS topics. 

Reviewer 2: I'm not sure `include_directories` is needed here.

Author: @bennowotny When I commented it out, CMake failed and gave me the error `pp:1:10: fatal error: ros/init.h: No such file or directory
    1 | #include "ros/init.h"`.

Reviewer 2: Ok, it can stay.  Sorry.

For the Python files, you will only need the `catkin_install_python` specification. This will list out the locations of the Python files being used.

In future projects, you might need to use other configurations, builds, and installs, but for the purposes of this training, you only need these for the example code to run.

You can also check out the package.xml file, which you also may need to tailor for future projects. However, everything should already be set up correctly, so do not change anything. At the top of the file, it should name the creator of the package. Then, it will include the license regarding open source policies. Right now, it should not specify a license because the license for the package for the WRoverPlayground training trickles down to this package. At the bottom, there should be various dependencies that allow us to use ROS and ROS topics, similar to the CMakeLists.txt `find_package`.

Now, we will actually run the program.

1. In the terminal, run the command `roscore`. 

2. In a new terminal window, change your current directory to be the catkin workspace. 

> In this case, use `cd WRoverPlayground`. 

Reviewer 2: Since we're in a container, you know/can specify the full path to the workspace:`/workspaces/WRoverPlayground`.  It'll be clearer for the students.

Author: @bennowotny When I use `cd /workspaces/WRoverPlayground`, it says that there is no such file or directory. Am I supposed to have a directory called `workspaces`?

Reviewer 2: This won't exist in your VM right now, but it will when Docker integration merges.

Author: Fixed in 09952ed.

3. Use the commands `catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=ON` and `source devel/setup.bash` in this order. 

> These commands adjust the catkin workspace based on the CMakeLists.txt and package.xml files. You must run these commands everytime you change these two files. 

Reviewer 2: VSCode and the Docker container will take care of these steps for the programmer.  Just use the built-in VSCode terminal

Reviewer 2: Maybe they can stay?  But I think they're redundant.

Author: @bennowotny So will students just need to use `rosrun`?

Reviewer 2: Ideally, yes.  The Docker container should take care of the setup automatically.

Author: Fixed in 6a86158 and bce99f6.

# WRoverPlayground - Topic Training
Now that you have completed the Tools Training, you are ready to learn about ROS topics. From the Setup Training, you should be familiar with:
- the basics of ROS Nodes
- `rosnode`
- `roscore`
- `rqt_graph`
- `rqt_plot`.

Once you complete this training, you should be familiar with:
- the purpose of topics
- the relationship between Publishers and Subscribers
- setting up Publishers and Subscribers
- running Publishers and Subscribers
- custom topic types
- and `rostopic type {topic}`.

## The Purpose of Topics
Though possible, it is not feasible to have one node contain all the information a rover would need. We need to spread out the information into different files. This is where ROS topics come in. The word "topic" sounds very general, but it is actually a specific concept that allows ROS nodes to communicate with each other. They are a way to stream data in and out of a program.

> Using topics, our code now becomes 'hot-swappable.' Topics do not care which node is sending or receiving data, so we can easily create a node that serves data from a physical sensor as well as a mock node to test other nodes without needing to change those nodes. The reason is because the interfaces provided through the published data versus an actual sensor are equivalent.

Reviewer: I'm a bit confused by this paragraph, especially the second sentence. Also, what do you mean by hot-swappable? 

I think this chapter would be more readable if you swap this paragraph with the one below. This way the flow of this chapter would be:
1. Introduce what topics do (stream data in between nodes)
2. Give an example of why we use them (to specialize nodes, such as the example you gave where we can have one node dedicated to receiving user input and another to write to the motor)
3. Emphasize the significance of this approach (this are hot-swappable, easier to maintain or change, etc.)

Author: Fixed in d3feca6.

## Publisher/Subscriber Relationship
The Publisher and Subscriber are both nodes that are connected by a topic, allowing them to send messages between each other. The Publisher node publishes messages on the topic, while the Subscriber node subscribes to the same topic to receive the Publisher's messages. 

Topics do not need to have a one-to-one relationship. For example, it is possible for a topic to have multiple Subscribers, which allows a single node to communicate with other nodes.

Also, Publishers and Subscribers do not need to be in separate files. In fact, if the Publisher and Subscriber rely on each other's data, it may be better to keep them in the same file. However, a node should not publish to itself as this would be incredibly inefficient. A way to do this properly would be to have the Subscriber receive data from one topic, the data to be manipulated in some way, and the Publisher to send the result on another topic. 

An example of a Publisher and Subscriber residing in the same file will be provided in the next section.

## Setting up a Publisher and Subscriber

> Within the directory /workspaces/WRoverPlayground/src/topic_training/src, you will find two files called Demo(.cpp or .py). These files will show a basic class containing both a Publisher and Subscriber. There will be comments throughout the files that explain the purpose of specific sections of the code.

Reviewer: Using `the code markdown` can make this file path more readable. 

```suggestion
Within the directory `/workspaces/WRoverPlayground/src/topic_training/src`, you will find two files called Demo(.cpp or .py). These files will show a basic class containing both a Publisher and Subscriber. There will be comments throughout the files that explain the purpose of specific sections of the code.
```

Author: Fixed in 5182f6b.

Note that there are a few sections of code that are marked as "Optional." You do not need to include everything in the examples when creating your own code. They are simply showing one way of setting up a Publisher and Subscriber in the same file.

> One file will be written in C++, and the other will be in Python. It is recommended that you choose the language you are most comfortable with to complete the training. Afterwards, you should discuss with one of the Software Leads to determine which part of the rover you would like to be assigned to. You will be given small tasks first, working your way up to more difficult projects. If these tasks/projects require a different language, then you can look through this training again or take what you know from the first time and learn as you go.

Reviewer: Clarify that new members will be assigned to projects after all training is complete.

```suggestion
One file will be written in C++, and the other will be in Python. It is recommended that you choose the language you are most comfortable with to complete the training. After you are done with all trainings in `WRoverPlayground`, you should discuss with one of the Software Leads to determine which part of the rover you would like to be assigned to. You will be given small tasks first, working your way up to more difficult projects. If these tasks/projects require a different language, then you can look through this training again or take what you know from the first time and learn as you go.
```

Author: Fixed in b38f9b8.

Conintue reading here and go into one of the Demo files based on the language you chose. Make sure you understand what is being said. The general idea of the Demo (.cpp or .py) file is that it takes in data from a separate Publisher file (PlotDemo.cpp or .py) as a float on the topic `plot` and sends that data in the form of a String on the topic `chatter`.

### C++

A basic setup for ROS topics in C++ would look something like this:

``` C++
ros::init(argc, argv, "Demo");
ros::NodeHandle nhandle;
ros::Subscriber sub = nhandle.subscribe("plot", QUEUE_SIZE, chatterCallback);
ros::Publisher pub = nhandle.advertise<std_msgs::String>("chatter", QUEUE_SIZE);
```

To break this down,

``` C++
ros::init(argc, argv, "Demo");
```
Initialize ROS and specify the name of the node (which is the name of the file in this case). This requires the parameters from the main function header (argc and argv).

``` C++
ros::NodeHandle nhandle;
```

> Create a handle for the node. This will do the actually initialization of the node in the next few lines.

Reviewer: Please correct me if I misremembered, but the node is initialized using `ros::init(...)`. Here the `NodeHandle` allows the programmer to create subscriber and published objects, which are more like "parts" or "functions" of a node.

Author: You're right. I worded it weirdly. I meant that new members will be using it to initialize the Subscriber and Publisher objects in the next couple of lines.

Author: Fixed in 30181b7.

``` C++
ros::Subscriber sub = nhandle.subscribe("plot", QUEUE_SIZE, chatterCallback);
```
Initialize the Subscriber. The subscribe() function is how we tell ROS we want to receive messages on a specific topic. State the name of the topic (`plot`), the queue size (1), and the function that will be called when a message is received (`chatterCallBack(msg)`). The queue size is the maximum number of messages kept as a buffer before throwing away old ones if we are publishing too quickly.

``` C++
ros::Publisher pub = nhandle.advertise<std_msgs::String>("chatter", QUEUE_SIZE);
```

> Initialize the Publisher. The advertise() function is how we tell ROS we want to publish on a topic. State the message type we will be publishing (std_msgs::String), the name of the topic (`chatter`), and the queue size (1).

Reviewer: > The advertise() function is how we tell ROS we want to publish on a topic.

This wording might confuse a new member into thinking that the `advertise()` function publishes a message to the topic when in reality it only creates a publisher object. Might be better to rephrase this.

```suggestion
Initialize the Publisher. The advertise() function creates a publisher object that we can use to publish messages to a topic later on. State the message type we will be publishing (std_msgs::String), the name of the topic (`chatter`), and the queue size (1).
```

Author: Fixed in 0be0e6b.

``` Python
rospy.Subscriber('plot', Float64, callback)
```
Initialize the Subscriber. State the name of the topic (`plot`), the message type we will be publishing (float, which is actually in the class std_msgs.msg.Float64), and the function that will be called when a message is received (callBack(data)).

``` Python
pub = rospy.Publisher('chatter', String, queue_size=QUEUE_SIZE)
```
Initialize the Publisher. State the name of the topic (`chatter`), the message type we will be publishing (String, which is actually in the class std_msgs.msg.String), and the queue size (1). The queue size is the maximum number of messages kept as a buffer before throwing away old ones if we are publishing too quickly.

The callback function for a Subscriber looks something like this:

``` Python
def callback(data: Float64) -> None:
    global pi 
    pi = data.data
```
This function will get called when a new message has arrived on the topic `plot`. The message received by the Subscriber, sent by the Publisher, is denoted as `msg`.

Publishing a message on a topic would look something like this:

``` Python
pi_str = "currently at %s" % pi

> pub.publish(hello_str)

Reviewer: The variable you are publishing here is a separate variable. Not sure if this is intentional or not.

Author: Fixed in bcee277.

```

To break this down,

``` Python
pi_str = "currently at %s" % pi
```
Create the message.

``` Python
pub.publish(hello_str)
```
Publish the message using the previously initialized Publisher `pub`.

If you have glanced at the C++ example code, you may have noticed the function `ros::spinOnce()` and its differences with `ros::spin()`. Python only has `rospy.spin()`, which can be completely replaced by a for loop, which is why you do not see it in the Python example code.

## Running a Publisher and Subscriber

> Before you are able to run the example code, find the CMakeLists.txt file in the directory /workspaces/WRoverPlayground/src/topic_training. This file will include the the configurations, builds, and installs required to run your program as well as the files you will be running. Do not change anything in the file as it should already be set up for the demo files to work. 

Reviewer: Same as before, adding `code` can make the path more readable and stand out. 

Also, it might be nice to highlight that they don't really need to understand `CMakeLists` as it's pretty complicated and I can imagine new members quitting because of how daunting it seems. 

```suggestion
Before you are able to run the example code, find the CMakeLists.txt file in the directory `/workspaces/WRoverPlayground/src/topic_training`. This file will include the the configurations, builds, and installs required to run your program as well as the files you will be running. Do not change anything in the file as it should already be set up for the demo files to work. You also don't need to understand anything in the file, as it's way beyond the scope of this training. 
```

Author: Fixed in 5182f6b and 59a2d15.

The `find_package`, `catkin_package`, and `include_directories` specifications are required for both the C++ and Python files to work. These provide the packages and directories needed for running ROS and ROS topics. 

> To run the C++ files, you will need the `add_executable`, `add_dependencies`, and `target_link_libraries` specifications. There will be three of each, representing the three C++ files in this section. `add_executable` determines which file you will use with the command `rosrun`, `add_dependencies` shows what the executable file depends on, and `target_link_libraries` links libraries or executable targets that may be used in the program. In larger projects, these specifications may contain different names, but since the code examples in this training are simple, they are all the same.

Reviewer: I really like what you wrote here about what to include in the `CMakeLists`. It's important we have this documented somewhere as it's definitely hard for one to figure it out themselves without any outside help. But I feel like this should probably go elsewhere, maybe in another chapter or even an entirely new training module. It diverts focus away from the purpose of this training module (learning how to use chapters), and introduces a lot of new keywords that aren't needed for students to complete this training.

Author: I can see how new members could get overwhelmed quickly if they think we expect them to completely understand the `CMakeLists` file. What should I do with it? I also wrote something for the `rqt_*` commands that Ben told me to remove and save since it should be placed somewhere else. What training should these two explanations be?

Reviewer: That's a great question, thanks for bringing this up.

@bennowotny I'm curious what's your opinion on this? The `rqt` related writeup can go in the Debug training, but what about `CMakeLists`? I feel like this could either be it's own training or as a standalone documentation somewhere in the repo.

Reviewer 2: The CMake Training should move to Setup. @co-li

Author: Fixed in a2e2adb.

While developing your code for the challenge, use rqt_graph and rqt_plot to make sure your nodes are working as expected.

## Defining a Custom Topic Type
A topic's type is determined by the type of the message being published. If you run the command `rostopic type chatter`, you should see the terminal print out "std_msgs/String" while roscore and rosrun are executing Demo(.cpp or .py). The command will print out "std_msgs/Float64" if you are executing PlotDemo(.cpp or .py).

A topic can have messages of type int8, int16, int32, int64, uint*, float32, float64, String, time, duration, other msg files, and array[] of variable length or fixed length. To message a different type, just change where ever there would be a String in Demo(.cpp or .py) to your desired type and update the includes (C++) or imports (Python) at the top of the file. Remember though, use message types instead of variable types, that is, use std_msgs/String (for C++) or std_msgs.msg.String (for Python) instead of just String.

Along with the types listed above, you can also create custom topic types. Custom topic types are defined in .msg files. 

> 1. In /workspaces/WRoverPlayground/src/topic-training/msg/, find the CustomType.msg file. 

Reviewer: Same as before, adding 'code' to make the file path more readable. 

```suggestion
1. In `/workspaces/WRoverPlayground/src/topic-training/msg/`, find the CustomType.msg file. 
```

Author: Fixed in 5182f6b.

Publishing a message on a topic would look something like this:

``` Python
pi_msg = "currently at %s" % pi
pub.publish(pi_msg)
```

To break this down,

``` Python
pi_msg = "currently at %s" % pi
```
Create the message.

``` Python
pub.publish(pi_msg)
```
Publish the message using the previously initialized Publisher `pub`.

If you have glanced at the C++ example code, you may have noticed the function `ros::spinOnce()` and its differences with `ros::spin()`. Python only has `rospy.spin()`, which can be completely replaced by a for loop, which is why you do not see it in the Python example code.

## Running a Publisher and Subscriber

> Before you are able to run the example code, find the CMakeLists.txt file in the directory `/workspaces/WRoverPlayground/src/topic_training`. This file will include the the configurations, builds, and installs required to run your program as well as the files you will be running. Do not change anything in the file as it should already be set up for the demo files to work. You also do not need to understand anything in the file as it is way beyond the scope of this training.

Reviewer 2: Rephrasing:
```suggestion
Before you are able to run the example code, find the CMakeLists.txt file in the directory `/workspaces/WRoverPlayground/src/topic_training`. This file will include the the configurations, builds, and installs required to run your program as well as the files you will be running. Do not change anything in the file as it should already be set up for the demo files to work. This file will be covered by the Setup training.
```

Author: Fixed in a2e2adb.
